spec/path_normalizer_spec.cr

```diff
@@ -0,0 +1,29 @@
+require "./spec_helper"
+
+describe LuckyRouter::PathNormalizer do
+  describe ".normalize" do
+    it "turns regular path parts into slash delimitted string" do
+      path_parts = LuckyRouter::PathPart.split_path("/api/v1/users")
+
+      result = LuckyRouter::PathNormalizer.normalize(path_parts)
+
+      result.should eq("/api/v1/users")
+    end
+
+    it "gives path variables a generic name" do
+      path_parts = LuckyRouter::PathPart.split_path("/users/:id")
+
+      result = LuckyRouter::PathNormalizer.normalize(path_parts)
+
+      result.should eq("/users/:path_variable")
+    end
+
+    it "removes question mark from optional path" do
+      path_parts = LuckyRouter::PathPart.split_path("/users/?name")
+
+      result = LuckyRouter::PathNormalizer.normalize(path_parts)
+
+      result.should eq("/users/name")
```

Comment:
Maybe this could be `/users/optional_variable`? I'm just wondering if `/users/?name` and `/users/?username` would conflict. It seems they wouldn't, but they probably should, right?

Reply:
I think you're getting mixed up, `?name` is not a path variable, it's a static path part. The actual request comes in as `/users/name`. Those two do conflict because we break each path with optional path parts into the base part without any optional and add that route first and then add each optional route back one by one.  So if you register a route as `/users/?name`, two routes are added to the internal lookup
- `/users`
- `/users/name`

After the paths are inserted, we no longer consider optional parts because we added individual lookups for each possible route the path can handle.

Reply:
Oh, I thought it was a variable so like `/users/?name` would match `/users` and `/users/jeremy` and `/users/matthew`. Does that have to be `/users/?:name` for that to work? Or do we not have that case?

Reply:
Yes, we handle that as well 👍

Reply:
Ok. Should we add a spec to make sure that's tested in all of this then?

Reply:
Good idea, and done!